### ISSUE TITLE
False Error About a Fragment Not Being Used (Same Query Works Fine in gatsby-node ... Just Not on a Page)

### ISSUE THREAD

**Issue by machineghost** [gatsbyjs/gatsby#26751]
## Description

I'm getting an erroneous error.  When I try to include a fragment in a page query, Gatsby breaks, _supposedly_ because that fragment isn't being used ...

>success onPreExtractQueries - 0.002s
>
>  ERROR #85901  GRAPHQL
> 
> There was an error in your GraphQL query:
> 
> Fragment "productFields" is never used.
> 
> GraphQL request:53:3
> 52 |   }
> 53 |   fragment productFields on PostGraphile_Product {
>    |   ^
> 54 |     title
> 
> File: src/pages/admin.js:60:3
> 
> failed extract queries from components - 0.990s

But the thing is ... that fragment _is_ being used (multiple times) in the query directly above it:

    export const query = graphql`
                # parts of query you don't care about
                products {
                  ...productFields
                }

Also, perhaps more importantly, I'm using the _exact same_ query (I literally copy/pasted it) in my `gatsby-node.js` file, and it works perfectly fine there.

Unless there is some prohibition against using fragments in page queries this seems like a bug.

### Steps to reproduce

My repo isn't public unfortunately, but I've included the relevant query bits, and I'd be happy to answer any questions.

### Expected result

What should happen?

I should be able to use the same exact query as an export on a page that I can already use (without the tag) in `gatsby-node.js`.

### Actual result

See above

### Environment

  System:
    OS: Linux 5.4 Linux Mint 20 (Ulyana)
    CPU: (8) x64 Intel(R) Core(TM) i7-4810MQ CPU @ 2.80GHz
    Shell: 5.0.17 - /bin/bash
  Binaries:
    Node: 14.8.0 - /usr/bin/node
    npm: 6.14.7 - /usr/bin/npm
  Browsers:
    Chrome: 84.0.4147.135
    Firefox: 79.0
  npmPackages:
    gatsby: ^2.24.15 => 2.24.15 
    gatsby-cli: ^2.12.47 => 2.12.47 
    gatsby-image: ^2.4.7 => 2.4.7 
    gatsby-plugin-import: ^2.1.5 => 2.1.5 
    gatsby-plugin-manifest: ^2.4.11 => 2.4.12 
    gatsby-plugin-modal-routing: ^1.2.0 => 1.2.0 
    gatsby-plugin-offline: ^3.2.9 => 3.2.11 
    gatsby-plugin-react-helmet: ^3.3.4 => 3.3.4 
    gatsby-plugin-root-import: ^2.0.5 => 2.0.5 
    gatsby-plugin-s3: ^0.3.4 => 0.3.4 
    gatsby-plugin-sharp: ^2.6.11 => 2.6.12 
    gatsby-source-filesystem: ^2.3.11 => 2.3.12 
    gatsby-source-pg: ^0.2.1 => 0.2.1 
    gatsby-transformer-sharp: ^2.5.5 => 2.5.5 
  npmGlobalPackages:
    gatsby: 2.24.15

**Comment by vladar** [comment]
Hi @machineghost 

I tried reproducing this on a simple blog project but it worked as expected. So it is incredibly helpful if you're able to create a [minimal reproduction][reproduction]. This is a simplified example of the issue that makes it clear and obvious what the issue is and how we can begin to debug it.

You can use the project I've started as a base for the repro: https://github.com/vladar/gh-26751

Thanks for using Gatsby! 💜

[reproduction]: https://gatsby.dev/reproduction

**Comment by machineghost** [comment]
That repo does:

```
export const query = graphql`
  {
    allPosts: allMdx {
      edges {
        node {
          ...productFields
        }
      }
    }
  }
`

export const fragment = graphql`
  fragment productFields on Mdx {
    id
    frontmatter {
      title
      coverImage {
        id
        name
      }
    }
  }
```

... or in other words, it's exporting a separate fragment constant.  But again, I'm doing both in a single string:

export const query = graphql`
  {
    allPosts: allMdx {
    #rest of query
    }
  }
  fragment productFields on Mdx {
    # ...
`

And again this *exact* same query works perfectly fine as a single string inside `gatsby-node.js`.  So maybe my real question is ... why prohibit the exact same thing you allow elsewhere?

**Comment by vladar** [comment]
I tried both options - export in a separate variable and define within the same query string. Both options worked for me, that's why we need a repro.

**Comment by machineghost** [comment]
I tried to re-create using the repo, but I wasn't able to.  To do so I'd have to basically re-create my entire database, after cleaning it up to just have the parts for this bug, because my GraphQL schema (using Postgraphile) depends on it.

Unfortunately that's more work than I want to do, especially since ... I figured out a workaround :)  And _hopefully_ that workaround can lead to a reproducible case.  In essence, I just changed:

    export const query = graphql`
      query QueryName {

to:

    export const query = graphql`
      query QueryName2 {

In other words, renaming my query made the "`Fragment "productFields" is never used.`" error go away.

Presumably this was only a problem because I also had a query named `QueryName` in my `gatsby-node.js`.  My guess is that somehow all queries are sharing a namespace.

Still, it feels like there's a bug here: I would have expected an error to the effect of "You can't redefine a query with the same name!" ... but I wound up with something about a fragment not being used?

**Comment by vladar** [comment]
This sounds very odd. Queries in `gatsby-node` shouldn't affect page queries in any way. So now if you rename your page query back to `query QueryName {` does it start failing again?

**Comment by machineghost** [comment]
Yup (but you may want to skip to the TLDR at the  end): I just tried it again, to "sanity check", and ...

>  ERROR #85901  GRAPHQL
> 
> There was an error in your GraphQL query:
> 
> Unknown fragment "productFields".
> 
> GraphQL request:33:22
> 32 |                 products {
> 33 |                   ...productFields
>    |                      ^
> 34 |                 }

To get further debugging info (since I'm being an ass and not giving you a repro case) I also tried fixing the name of the query in `gatsby-node.js`, assuming I'd get the same result ... but I **didn't**, I got the same error!  (And to be clear, I'm completely restarting thet Gatsby/Webpack server every time.)

So then I tried changing the *fragment* name (`productFields2`), and ... same error.  This sort of shot down my namespace theory ... or so I thought.  But then I did a quick "find", and realized I *also* had a (nascent version) of that query on yet another page, and crucially it had the same name (although, being nascent, it wasn't really the same query: it didn't use the fragment).

Next I tried replacing the query in that file with the one in the first file, and finally I got ... well I got an error, but this one at least made sense:

> Multiple "root" queries found: "GetGenres" and "GetGenres".
> Only the first ("GetGenres") will be registered.
> 
> Instead of:
> 
>    1 | query GetGenres {
>    2 |   postgres {
>    3 |     #...
>    4 |   }
>    5 | }
>    6 | 
>    7 | query GetGenres {
>    8 |   postgres {
>    9 |     #...
>   10 |   }
>   11 | }
> 

**TLDR**
Based on my observations (and again sorry for just giving observations and not a repro), the issue *seems to be* caused by having two queries in two different pages (I had `index.js` and `admin.js`), with the same names, but one of them uses fragments and the other doesn't.

Under those conditions (and possibly still others that I haven't noticed), it seems the correct error ("Multiple "root" queries found") *isn't* thrown ... instead you get this weird error about fragments being unused.  And actually that *kind of* makes sense: if query #1 doesn't use the fragment, and it prevents query #2 with the same name that does use the fragment from being loaded (but the fragment from query #2 *does* get loaded) ... then you wind up with query #1 and the fragment from query #2, and that would (in a computery way) justify saying the fragment is unused.

**Comment by vladar** [comment]
@machineghost Excellent investigation, thanks! So this is a very interesting edge case. We should definitely output a correct error about duplicate query names first.

If someone is willing to submit a PR with a fix - we'll be happy to accept it! If nobody is up to this, I will do it myself when I have a chance (but no ETA as it has a relatively low impact)

**Comment by machineghost** [comment]
If you can point me in the general direction of the code involved I _might_ take a shot ... but no promises.

**Comment by vladar** [comment]
I think we can have a check for query duplicates in this function:

https://github.com/gatsbyjs/gatsby/blob/733fabf0cdcc6beba8d454b771577ac38e59fa25/packages/gatsby/src/query/query-compiler.js#L195

We already have validation for fragment uniqueness there. But we treat queries and fragments differently. Basically here is a relevant piece:

https://github.com/gatsbyjs/gatsby/blob/733fabf0cdcc6beba8d454b771577ac38e59fa25/packages/gatsby/src/query/query-compiler.js#L229-L260

The `if (def.kind === Kind.OPERATION_DEFINITION) {` should contain a very similar logic as the fragments branch below.

Technically we do this validation for queries in another function: `processDefinitions` (I am not entirely sure why though):

https://github.com/gatsbyjs/gatsby/blob/733fabf0cdcc6beba8d454b771577ac38e59fa25/packages/gatsby/src/query/query-compiler.js#L302-L319

So the idea is to move it from `processDefinitions` to `extractOperations`. This will be also consistent with fragment duplicates validation.

**Comment by LekoArts** [comment]
Hi!

I'm closing this as a stale issue as in the meantime Gatsby 4 and related packages were released. You can check our [Framework Version Support Page](https://www.gatsbyjs.com/docs/reference/release-notes/gatsby-version-support/) to see which versions currently receive active support. If this is a feature request, please [create a discussion](https://github.com/gatsbyjs/gatsby/discussions/categories/ideas-feature-requests) as we moved feature requests from issues to GitHub Discussions.

Please try the mentioned issue on the latest version (using the `next` tag) and if you still see this problem, open a new bug report. It **must** include a [minimal reproduction](https://gatsby.dev/reproduction).

Thanks!

**Comment by rarest-adam** [comment]
I can confirm this bug is still happening in Gatsby 5.0.1. Changing the query name resolved the "Fragment is never used" for me too.

**Comment by chuongnh3atgmaildotcom** [comment]
Still happening in Gatsby 5.4.2. Use a different query name solved this.

**Comment by NeuroNine** [comment]
Can confirm this still occurs in 5.13.4. Ensuring that each of my queries have a unique name is still a valid workaround that resolves the issue.

**Comment by Toxicom** [comment]
5.14.5 still happening, the fix was to change all of my `query PageBySlug { ... }` to unique names instead of `PageBySlug`
